### PR TITLE
Refactor of cluster generators with dependency isolation

### PR DESCRIPTION
**Pull request by nareddyt**
This is a refactor of [cluster_generator.go](https://github.com/GoogleCloudPlatform/esp-v2/blob/c4796ebe6529da6911a7602ddce5295b8b0acd75/src/go/configgenerator/cluster_generator.go#L34) to split all the cluster creation functions into their own `ClusterGenerators`. Each `ClusterGenerator` can be created directly via the factory function `NewXYZFromServiceConfig(...)` instead of via `ServiceInfo`.

This PR only copies pre-existing cluster configs and parameterizes them for dependency isolation. No behavior change yet. Followup PR to:

1) Add implementation of factory functions
2) Add unit tests
3) Replace `cluster_generator.go` with this dependency injection logic.

Ref: [go/cloud-esf-config-gen-inputs](http://goto.google.com/cloud-esf-config-gen-inputs)

### CODE REVIEW

**Comment by nareddyt** [comment]
PTAL. This PR only adds new files, no risk

File: src/go/configgenerator/clustergen/helpers/backend.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	ClusterConnectTimeout  time.Duration
+	MaxRequestsThreshold   int
+	BackendDnsLookupFamily string
+
+	DNS *ClusterDNSConfiger
+	TLS *ClusterTLSConfiger
+}
+
+// GetName implements the ClusterGenerator interface.
```
Comment (TAOXUY) [inline src/go/configgenerator/clustergen/helpers/backend.go:33]: remove? As it doesn't need to implement ClusterGenerator. Its owner will be responsible for that.

File: src/go/configgenerator/clustergen/helpers/backend.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
+package helpers
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/clustergen/helpers/backend.go]: Done

File: src/go/configgenerator/clustergen/helpers/backend.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"google.golang.org/protobuf/types/known/durationpb"
+	wrappers "google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// BackendCluster is an Envoy cluster to communicate with a backend.
+//
+// This should NOT be used directly.
+// Use it via an abstraction like RemoteBackendCluster or LocalBackendCluster.
+type BackendCluster struct {
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/clustergen/helpers/backend.go]: Done

File: src/go/configgenerator/clustergen/helpers/backend.go (hunk 1 above)
Comment (nareddyt) [reply src/go/configgenerator/clustergen/helpers/backend.go]: Sure, done

File: src/go/configgenerator/clustergen/helpers/backend.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	DNS *ClusterDNSConfiger
+	TLS *ClusterTLSConfiger
+}
+
+// GetName implements the ClusterGenerator interface.
+func (c *BackendCluster) GetName() string {
+	return c.ClusterName
+}
+
+// GenConfig implements the ClusterGenerator interface.
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/clustergen/helpers/backend.go]: Done

**Review by shuoyang2016 — APPROVED** [review]

**Pull request merged**